feat: Add OAuth support for Tool Server authentication

jescalada opened the pull request:
This PR adds OAuth support for Tool Server authentication.

**Note:** This PR enables logging in through a provider configured in the backend, and getting an "access token" which can be used to validate subsequent requests. Available providers are Google and Atlassian. Feel free to extend this for the providers that are most relevant!

This PR **does not** implement the entire OpenWebUI -> OAuth + MCP flow discussed in #14121. I hit a few roadblocks along the way and I believe some extra refactoring work is needed to get it to work. I'll write some suggestions about how we could approach that and some clues I found.

# Changelog Entry

### Description

- This PR enables logging in through a provider configured in the backend, and getting an "access token" which can be used to validate subsequent requests. Available providers are Google and Atlassian. This can be easily extended with any provider needed.

### Added

- Added OAuth authentication option for Tool Servers
- Added Tool Server provider configuration variables and provider loading
- Added relevant endpoints for handling OAuth login for tool servers
- Extended `AddServerModal` to show providers and handle entire OAuth flow, including **manual** token refresh

---

### Additional Information

- Relevant issues
  - #14121
  - #13457

### Screenshots or Videos

#### Demo Video
[Demo](https://github.com/user-attachments/assets/b992d959-a7d4-445e-896a-b872c896fb9f)

Note that my cursor isn't showing in the video, sorry about that!

### Testing

If you'd like to test this PR, you can create your own Google auth client here: https://console.cloud.google.com/apis/credentials

Otherwise, you can do the same with an Atlassian OAuth 2.0 app: https://developer.atlassian.com/console/myapps

Otherwise, I can post my credentials and then manually add a reviewer's email (otherwise it won't work, since it's a test auth app).

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

jescalada committed:
enh: add TOOL_GOOGLE configs

jescalada committed:
enh: add TOOL_ATLASSIAN configs

jescalada committed:
enh: add TOOL_SERVER_OAUTH_PROVIDERS loader

jescalada committed:
fix: extract tool server config scopes

jescalada committed:
feat: add ToolServerOAuthManager class and init

jescalada committed:
feat: add tool server login and callback support for OAuth

jescalada committed:
enh: import ToolServerOAuthManager to main.py

jescalada committed:
feat: add login endpoint for any OAuth provider

jescalada committed:
feat: add OAuth callback endpoint

jescalada committed:
feat: add endpoint and middleware for fetching providers

jescalada committed:
enh: add event listener for OAuth popup

jescalada committed:
enh: add OAuth flow handler and verification check

jescalada committed:
feat: add OAuth login inputs and dynamic provider fetching

limbik-martinj commented:
Just a question:
How would this PR effect the following usecase?

I have a keycloak OAuth Server deployed and the open-webui is connected with OpenID Oauth.
My OpenAPI server tools also look for a Bearer access token, same one that the frontend webui just acquired.
I have a  customized version that allows me to pass the access token from the webui to the tool server.

Would this PR require me to login separately to aquire a separate token, that's not the same from the main login screen?

jescalada committed:
Update backend/open_webui/utils/oauth.py

Co-authored-by: Antti Kaihola <13725+akaihola@users.noreply.github.com>

jescalada commented:
### Update

I just updated this PR to complete the OAuth flow. I created this [Tool Server Demo repo](https://github.com/jescalada/tool-server-demo) to test that everything works.

![image](https://github.com/user-attachments/assets/a00cb4d9-7541-4246-bb28-b804602efacd)

### Important Note

Now, the OAuth flow extracts **two** important auth variables: `id_token`, which is **required** by OAuth flows for identifying the user, and `access_token` which is **optional** and provides access to a specific service or API.

![image](https://github.com/user-attachments/assets/a761414a-a398-4ecc-9a70-93ced43fd733)

jescalada committed:
fix: improve atlassian tool server configurability

jescalada committed:
feat: add checks for oauth auth_type to tool_server calls

jescalada committed:
feat: extract access_token/id_token and process in AddServerModal

jescalada committed:
fix: AddServerModal oauth handler refresh bug

jescalada committed:
feat: plug oauth tokens into tool server functions

jescalada committed:
feat: show both tokens in Connection modal

Classic298 commented:
### **WARNING:**
**You are trying to merge into `main`!**
Dangerous! This might get your PR closed without a comment or worst case: gets your changes accidentally merged into production!

jescalada commented:
Sorry for the slow update - I'll revisit this soon!

I misunderstood what the typical use case would look like, and the Python library I used as an example required **both** the `id_token` and the `access_token` in order to authenticate. There are some use cases where we want to figure out who the user is (`id_token`) and then validate that they can access a certain resource (`access_token`) in one go.

As you mentioned, the typical use case only needs an `access_token` to reach an API endpoint (in our case a tool server API).

I'll be fixing this sometime in the next week - mostly to revert the extra `id_token` stuff (the bug fixes and main flow should still work).

jescalada committed:
fix: remove usages of id_token for tool server auth

jescalada commented:
I just fixed the implementation to only use the `access_token` - the `id_token` stuff was a misunderstanding on my part. As you mentioned, the `access_token` is the [best practice for accessing an API resource](https://auth0.com/blog/id-token-access-token-what-is-the-difference/).

I also fixed up the [sample Python tool server](https://github.com/jescalada/tool-server-demo) so that it doesn't need the `id_token` anymore (in fact, it removes the JWT library entirely). Meanwhile, the user's identity is fetched through the `/userinfo` endpoint instead.

The fixed flow should be working as expected now:

<img width="2260" height="1314" alt="image" src="https://github.com/user-attachments/assets/c533167a-4753-48db-822b-97eb8085e2df" />

Would love a quick look at this @tjbck, let me know if anything in the sample flow isn't clear. Note that this does not set up the MCP support, but it's a step in the right direction since many users require simpler auth flows for their Tool Servers.

Thank you!

jescalada committed:
Merge branch 'dev' into add-tool-server-oauth

jescalada committed:
fix: token refresh not updating connection list

tjbck commented:
Closing in favour of OAuth2.1, OAuth2.1 support is implemented in such a way it could theoretically be extended to Tool servers as well.
